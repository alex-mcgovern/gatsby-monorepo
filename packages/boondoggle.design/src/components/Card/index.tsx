import type { ElementType, ReactElement, ReactNode } from "react";
import React, { forwardRef } from "react";
import { extractAtomsFromProps } from "@dessert-box/core";
import clsx from "clsx";
import type { GetSprinklesArgs } from "../../styles/getSprinkles.css";
import { getSprinkles } from "../../styles/getSprinkles.css";
import type {
  PolymorphicComponentPropWithRef,
  PolymorphicRef,
} from "../../types";
import { Box } from "../Box";
import * as styles from "./index.css";

type BaseCardProps<TPolymorphicAs extends ElementType> = GetSprinklesArgs &
  PolymorphicComponentPropWithRef<
    TPolymorphicAs,
    {
      children: ReactNode | Array<ReactNode>;
      className?: string;
    }
  >;

export type CardProps = <TPolymorphicAs extends ElementType = "button">(
  props: BaseCardProps<TPolymorphicAs>
) => ReactElement | null;

export const Card: CardProps = forwardRef(
  <TPolymorphicAs extends ElementType = "div">(
    {
      children,
      as,
      className: userClassName,
      ...rest
    }: BaseCardProps<TPolymorphicAs>,
    ref?: PolymorphicRef<TPolymorphicAs>
  ) => {
    /** Separate `GetSprinklesArgs` from other spread props, so we don't break Vanilla Extract */
    const { atomProps, otherProps } = extractAtomsFromProps(rest, getSprinkles);

    const Component = as || Box;

    const classNames = clsx(
      styles.cardStyle,
      getSprinkles(atomProps),
      userClassName
    );

    return (
      <Component
        {...{
          ref,
          className: classNames,
          ...otherProps,
        }}
      >
        {children}
      </Component>
    );
  }
);
