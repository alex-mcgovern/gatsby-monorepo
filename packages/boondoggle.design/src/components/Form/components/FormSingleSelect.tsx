import React, { useCallback } from "react";
import { useController, useFormContext } from "react-hook-form";
import type { SelectSingleProps } from "../../Select/SelectSingle";
import { SelectSingle } from "../../Select/SelectSingle";
import type { DropdownItem } from "../../Select/types";

export interface FormSingleSelectProps extends SelectSingleProps {
  errorMessage: string;
}

/** React Hook Form connected version of `SelectSingle`. Uses `useFormContext`
 * to access Hook Form's methods so can be deeply nested. */
export function FormSingleSelect({
  name,

  ...rest
}: FormSingleSelectProps) {
  const { control } = useFormContext();

  const {
    field: { onChange, ref },
    fieldState: { error },
  } = useController({
    name,
    control,
    rules: { required: true },
    defaultValue: "",
  });

  const handleChange = useCallback(
    (item: DropdownItem) => {
      return onChange(item.value);
    },
    [onChange]
  );

  return (
    <SelectSingle
      ref={ref}
      invalid={!!error}
      onValueChange={handleChange}
      name={name}
      {...rest}
    />
  );
}
