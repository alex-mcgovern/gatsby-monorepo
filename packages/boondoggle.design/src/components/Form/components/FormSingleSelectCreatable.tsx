import React, { useCallback } from "react";
import { useController, useFormContext } from "react-hook-form";
import type { SelectSingleCreatableProps } from "../../Select/SelectSingleCreatable";
import { SelectSingleCreatable } from "../../Select/SelectSingleCreatable";
import type { DropdownItem } from "../../Select/types";

export interface FormSingleSelectSingleCreatableProps
  extends SelectSingleCreatableProps {
  errorMessage: string;
}

/** React Hook Form connected version of `SelectSingle`. Uses `useFormContext`
 * to access Hook Form's methods so can be deeply nested. */
export function FormSingleSelectCreatable({
  name,

  ...rest
}: FormSingleSelectSingleCreatableProps) {
  const { control } = useFormContext();

  const {
    field: { onChange, ref },
    fieldState: { error },
  } = useController({
    name,
    control,
    rules: { required: true },
    defaultValue: "",
  });

  const handleChange = useCallback(
    (item: DropdownItem) => {
      return onChange(item?.value);
    },
    [onChange]
  );

  return (
    <SelectSingleCreatable
      ref={ref}
      invalid={!!error}
      onValueChange={handleChange}
      name={name}
      {...rest}
    />
  );
}
