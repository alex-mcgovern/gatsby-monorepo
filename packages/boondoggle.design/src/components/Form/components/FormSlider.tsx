import React from "react";
import type { Validate } from "react-hook-form";
import { useController, useFormContext } from "react-hook-form";
import type { SliderProps } from "../../Slider";
import { Slider } from "../../Slider";

export interface FormSliderProps extends SliderProps {
  /** Message to render when erroring. */
  errorMessage: string;
  /** Callback for validation, else simply validates is non-empty. */
  validateFunction?: Validate<string>;
  /** a11y label passed to `Label` component */
  label: string;
}

/**
 * React Hook Form connected version of Boondoggle's `Input`. Uses `useFormContext`
 * to access Hook Form's methods so can be nested in markup. Must be a descendant of `FormProvider`
 */
export function FormSlider({
  name,
  errorMessage,
  required,
  validateFunction,

  ...rest
}: FormSliderProps) {
  const { control } = useFormContext();

  const {
    field: { onChange, onBlur, ref },
    fieldState: { error },
  } = useController({
    name,
    control,
    rules: {
      // value,
      required: required && errorMessage,
      validate: (value) => {
        if (validateFunction) {
          return validateFunction(value) || errorMessage;
        }
        return !!value;
      },
    },
    defaultValue: "",
  });

  return (
    <Slider
      onChange={onChange}
      onBlur={onBlur}
      invalid={!!error}
      aria-required={required}
      errorMessage={errorMessage}
      name={name}
      // value={controlled  Value}
      ref={ref}
      {...rest}
    />
  );
}
