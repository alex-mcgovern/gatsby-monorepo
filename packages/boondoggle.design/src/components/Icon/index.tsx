import React from "react";
import { extractAtomsFromProps } from "@dessert-box/core";
import type { IconProp } from "@fortawesome/fontawesome-svg-core";
import type { FontAwesomeIconProps } from "@fortawesome/react-fontawesome";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import clsx from "clsx";
import type { GetSprinklesArgs } from "../../styles/getSprinkles.css";
import { getSprinkles } from "../../styles/getSprinkles.css";
import { Box } from "../Box";
import * as styles from "./icon.css";

export interface IconProps
  extends GetSprinklesArgs,
    Omit<
      FontAwesomeIconProps,
      | "border"
      | "color"
      | "display"
      | "fontSize"
      | "fontStyle"
      | "fontWeight"
      | "height"
      | "overflow"
      | "textDecoration"
      | "width"
    > {
  icon: IconProp;
}

export function Icon({ icon, className, ...rest }: IconProps) {
  const { atomProps, otherProps } = extractAtomsFromProps(rest, getSprinkles);

  return (
    <Box {...atomProps}>
      <FontAwesomeIcon
        icon={icon}
        className={clsx(styles.icon, className)}
        {...otherProps}
      />
    </Box>
  );
}
