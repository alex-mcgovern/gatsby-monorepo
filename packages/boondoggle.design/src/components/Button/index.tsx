/**
 * Approach adapted from a LogRocket blog by Ohans Emmanuel
 * https://blog.logrocket.com/build-strongly-typed-polymorphic-components-react-typescript/
 */
import type {
  ComponentPropsWithoutRef,
  ElementType,
  ReactElement,
} from "react";
import React, { forwardRef } from "react";
import { extractAtomsFromProps } from "@dessert-box/core";
import clsx from "clsx";
import type { VariantInteractiveElementSizeEnum } from "../../styles/common/variant_interactive_element_size.css";
import type { GetSprinklesArgs } from "../../styles/getSprinkles.css";
import { getSprinkles } from "../../styles/getSprinkles.css";
import type {
  PolymorphicComponentPropWithRef,
  PolymorphicRef,
} from "../../types";
import { Icon } from "../Icon/index";
import type { IconProps } from "../Icon/index";
import type { VariantButtonColorEnum } from "./button-color.css";
import { buttonTheme } from "./button.theme.css";
import { getButtonStyles, iconStyle } from "./index.css";
import type { VariantButtonAppearanceEnum } from "./variantButtonAppearance.css";

type BaseButtonProps<TPolymorphicAs extends ElementType> = Omit<
  GetSprinklesArgs,
  "color"
> &
  PolymorphicComponentPropWithRef<
    TPolymorphicAs,
    {
      /** Is button disabled. Mapped to html5 <button> `disabled` attribute and `aria-disabled` attribute. */
      disabled?: boolean;
      /** FontAwesome icon shown on the left side of button. */
      iconLeading?: IconProps["icon"];
      /** Props for leading icon */
      iconLeadingProps?: Omit<IconProps, "icon">;
      /** FontAwesome icon shown on the right side of button. */
      iconTrailing?: IconProps["icon"];
      /** Props for trailing icon */
      iconTrailingProps?: Omit<IconProps, "icon">;
      /** Title for button, shown in the UI */
      name: string;
      /** HTML button type, defaults to `button`. */
      type?: "button" | "submit" | "reset";
      /** Size of the button element */
      size?: VariantInteractiveElementSizeEnum;
      /** Color for the button */
      color?: VariantButtonColorEnum;
      /** Appearance for the button */
      appearance?: VariantButtonAppearanceEnum;
    }
  >;

type ButtonComponent = <TPolymorphicAs extends ElementType = "button">(
  props: BaseButtonProps<TPolymorphicAs>
) => ReactElement | null;

export type ButtonProps = ComponentPropsWithoutRef<typeof Button>;

/** -----------------------------------------------------------------------------
 * Button component
 * ------------------------------------------------------------------------------- */

export const Button: ButtonComponent = forwardRef(
  <TPolymorphicAs extends ElementType = "span">(
    {
      appearance = "primary",
      as,
      children,
      className: userClassName,
      color,
      disabled,
      iconLeading,
      iconLeadingProps,
      iconTrailing,
      iconTrailingProps,
      id,
      size = "md",
      type = "button",
      ...rest
    }: BaseButtonProps<TPolymorphicAs>,
    ref?: PolymorphicRef<TPolymorphicAs>
  ) => {
    /** Separate `GetSprinklesArgs` from other spread props, so we don't break Vanilla Extract */
    const { atomProps, otherProps } = extractAtomsFromProps(rest, getSprinkles);

    const Component = as || "button";

    const buttonClassNames = clsx(
      buttonTheme,
      getButtonStyles({ appearance, color, size }),
      getSprinkles(atomProps),
      userClassName
    );

    return (
      <Component
        {...{
          "aria-disabled": disabled,
          className: buttonClassNames,
          disabled,
          id,
          ref,
          type,
          ...otherProps,
        }}
      >
        {iconLeading && (
          <Icon
            className={iconStyle}
            icon={iconLeading}
            {...iconLeadingProps}
          />
        )}
        {children}
        {iconTrailing && (
          <Icon
            className={iconStyle}
            icon={iconTrailing}
            {...iconTrailingProps}
          />
        )}
      </Component>
    );
  }
);
