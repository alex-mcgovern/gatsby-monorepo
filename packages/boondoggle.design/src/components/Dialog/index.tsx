import type { ReactNode } from "react";
import React from "react";
import { faTimes } from "@fortawesome/free-solid-svg-icons";
import * as DialogPrimitive from "@radix-ui/react-dialog";
import { Box } from "../Box";
import { Button } from "../Button";
import * as styles from "./dialog.css";

export interface DialogProps {
  /** Element to use as Dialog trigger. Note: Must accept a ref. */
  triggerNode: ReactNode;
  /** Accessible title for dialog */
  title: string;
  /** Accessible description for dialog */
  description: string;
  /** Dialog content */
  children: ReactNode | Array<ReactNode>;
  callbackOnOpenChange: (isOpen: boolean) => void;
  isOpen: boolean;
}

export function Dialog({
  triggerNode,
  title,
  description,
  children,
  isOpen,
  callbackOnOpenChange,
}: DialogProps) {
  return (
    <DialogPrimitive.Root onOpenChange={callbackOnOpenChange} open={isOpen}>
      {/**
       * Allow custom trigger node, e.g. menu icon.
       * ToDo: Figure out a way to require triggerNode to accept ref, or to
       * wrap triggerNode so it is always able to accept a ref.
       */}
      <DialogPrimitive.Trigger asChild>{triggerNode}</DialogPrimitive.Trigger>

      <DialogPrimitive.Portal>
        <DialogPrimitive.Overlay className={styles.dialogOverlay} />
        <DialogPrimitive.Content className={styles.dialogContent}>
          {/* ---------------------------------------------——
           * DIALOG HEADER
           * ---------------------------------------------—— */}
          <Box
            display="flex"
            justifyContent="space-between"
            alignItems="center"
          >
            <DialogPrimitive.Title asChild>
              <Box as="h2" fontStyle="h4" marginY="none">
                {title}
              </Box>
            </DialogPrimitive.Title>

            <DialogPrimitive.Close asChild>
              <Button
                size="md_square"
                appearance="secondary"
                iconLeading={faTimes}
                aria-label="Close"
                type="submit"
                width="100%"
              />
            </DialogPrimitive.Close>
          </Box>

          <DialogPrimitive.Description>
            {description}
          </DialogPrimitive.Description>

          {/* ---------------------------------------------——
           * DIALOG CONTENT
           * ---------------------------------------------—— */}

          {children}
        </DialogPrimitive.Content>
      </DialogPrimitive.Portal>
    </DialogPrimitive.Root>
  );
}

export const DialogCloseButton = DialogPrimitive.Close;
