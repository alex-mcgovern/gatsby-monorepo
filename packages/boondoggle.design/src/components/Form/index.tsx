import type { FormEvent, ReactElement } from "react";
import React, { useCallback, useEffect, useMemo } from "react";
import { FormProvider, useForm } from "react-hook-form";
import { Box } from "../Box";
import { Button } from "../Button";
import type { FormInputProps } from "./components/FormInput";
import type { FormSingleSelectProps } from "./components/FormSingleSelect";
import type { FormSingleSelectSingleCreatableProps } from "./components/FormSingleSelectCreatable";
import { getHookFormButtonIconProps } from "./utils/getHookFormButtonIcon";

export interface FormProps {
  callbackOnSuccessfulFormSubmission: () => void;
  children: Array<
    | ReactElement<FormInputProps>
    | ReactElement<FormSingleSelectSingleCreatableProps>
    | ReactElement<FormSingleSelectProps>
  >;
  /** ToDo(react-hook-form): Fix submission handler types */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  handleFormSubmission: (...args: Array<any>) => Promise<unknown>;
  submitButtonText: string;
  disabled?: boolean;
}

export function Form({
  children,
  callbackOnSuccessfulFormSubmission,
  handleFormSubmission,
  submitButtonText,
  disabled,
}: FormProps) {
  /** ---------------------------------------------
   * Initialise react-hook-form and subscribe to form state proxy
   * https://react-hook-form.com/api/useform/formstate
   * ----------------------------------------------- */

  const reactHookFormMethods = useForm();

  const { formState, handleSubmit, reset } = reactHookFormMethods;

  const {
    isDirty,
    isSubmitSuccessful,
    isSubmitted,
    isSubmitting,
    isValid,
    isValidating,
    errors,
  } = formState;

  /** ---------------------------------------------
   * Update icon based on form state. (Add a bit of ✨fun✨)
   * ----------------------------------------------- */

  const { buttonIcon, buttonIconProps } = useMemo(() => {
    return getHookFormButtonIconProps({
      isValid,
      isValidating,
      isSubmitting,
      isSubmitSuccessful,
      isSubmitted,
      isDirty,
      errors,
    });
  }, [
    isDirty,
    isSubmitSuccessful,
    isSubmitted,
    isSubmitting,
    isValid,
    isValidating,
    errors,
  ]);

  /** ---------------------------------------------
   * Handlers for submission
   * ----------------------------------------------- */

  /**
   * Submission handler passed from parent scope
   */
  const onSubmit = useCallback(
    async (event: FormEvent) => {
      return handleSubmit(handleFormSubmission)(event);
    },
    [handleFormSubmission, handleSubmit]
  );

  /**
   * If the function passed to and called with `handleSubmit` resolves, `isSubmitSuccessful`
   * will trigger the `useEffect` and we can notify the parent component about the form submission.
   *
   * Note: because of the way react-hook-form batches updates to `formState`, it is  better to attach
   * `isSubmitSuccessful` as a dependency to a `useEffect`, rather than chain a `.then()` off the form handler.
   */
  useEffect(() => {
    if (isSubmitSuccessful) {
      reset();
      if (callbackOnSuccessfulFormSubmission) {
        callbackOnSuccessfulFormSubmission();
      }
    }
  }, [reset, isSubmitSuccessful, callbackOnSuccessfulFormSubmission]);

  /** -----------------------------------------------------------------------------
   * Markup
   * ------------------------------------------------------------------------------- */

  return (
    <FormProvider {...reactHookFormMethods}>
      <Box as="form" onSubmit={onSubmit} display="grid" gap="spacing2">
        {children}

        <Button
          aria-label={submitButtonText}
          size="lg"
          width="100%"
          type="submit"
          iconTrailing={buttonIcon}
          iconTrailingProps={buttonIconProps}
          disabled={disabled}
        >
          {submitButtonText}
        </Button>
      </Box>
    </FormProvider>
  );
}
