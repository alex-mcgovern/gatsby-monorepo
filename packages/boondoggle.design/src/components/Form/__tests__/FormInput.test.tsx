/** @jest-environment jsdom */
import React from "react";
import { cleanup } from "@testing-library/react";
import { renderTestComponent } from "../../../../../../test/renderTestComponent";
import { FormInput, FormInputProps } from "../components/FormInput";
import { FormTestProvider } from "../components/TestFormProvider";

afterEach(cleanup);

/**
 * UPDATING VALUE
 */

test("Given required props, renders without errors", async () => {
  const { component } = renderTestComponent<FormInputProps>(
    <FormTestProvider>
      <FormInput
        id="input"
        name="Test input"
        errorMessage="Error"
        label="test input"
      />
    </FormTestProvider>
  );

  expect(component as HTMLInputElement).not.toBeNull();
});

test("Given required props, matches snapshot", async () => {
  const { component } = renderTestComponent<FormInputProps>(
    <FormTestProvider>
      <FormInput
        id="input"
        name="Test input"
        errorMessage="Error"
        label="test input"
      />
    </FormTestProvider>
  );

  expect(component as HTMLInputElement).toMatchSnapshot();
});

test("Given an input without value, when changing, it should have the new value", async () => {
  const { user, getByRole } = renderTestComponent<FormInputProps>(
    <FormTestProvider>
      <FormInput
        id="input"
        name="Test input"
        errorMessage="Error"
        label="test input"
      />
    </FormTestProvider>
  );

  const input = getByRole("textbox") as HTMLInputElement;

  expect(input?.value).toBe("");
  await user.type(input, "New value");
  expect(input?.value).toBe("New value");
});
