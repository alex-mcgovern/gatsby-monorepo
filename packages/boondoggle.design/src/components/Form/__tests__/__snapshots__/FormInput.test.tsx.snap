// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given required props, matches snapshot 1`] = `
<input
  aria-invalid="false"
  class="Input_inputElement__17v4xld9 Input__17v4xld8 getSprinkles_paddingLeft_spacing1__1jganc99g getSprinkles_paddingRight_spacing1__1jganc99v getSprinkles_width_100%_mobile__1jganc9f9 getSprinkles_height_100%_mobile__1jganc9ho"
  data-testid="tested-component"
  id="input"
  name="Test input"
  value=""
/>
`;
